hedyorg/hedy
# 🪲 Prevent stopped programs from executing

In Skulpt there really isn't a way to stop the execution of a program. If the program uses things like prompts for the `input` function, or sleep, then those promises are still somewhere in the JavaScript handler. Therefore, when you execute a new program while another one is still waiting for a promise to be fulfilled (i.e: waiting for user input through an ask prompt), when the new program finishes execution, the old one will continue its path.

To fix this I added a new field to `sessionStorage` which will have an id for the program being executed, and I edited Skulpt to pass the id of the program. Therefore, when an old program is being executed, its id will be different than the current one and therefore will finish its execution.


Fixes #5211

**How to test**

* Go to level 14 and execute the TicTacToe adventure, fill the first ask.
* With an ask prompt still open, switch to the Quizmaster tab and execute the program.
* Only the Quizmaster program should be executed.
